tests/apps/test_cost.py

```diff
@@ -91,7 +94,288 @@ def test_grad_threshold(self, k):
+
+        cost_fn = train.Stochastic(h, vgbs)
+        h0 = cost_fn.h_reparametrized(zeros[0], params)
+        h1 = cost_fn.h_reparametrized(ones[0], params)
+        eval_expected = (h0 + h1) * 0.5
+
+        eval = cost_fn(params, 2 * n_samples)
+        assert np.allclose(eval, eval_expected)
+
+    @pytest.mark.parametrize("threshold", [True, False])
+    def test_gradient_one_sample(self, vgbs, dim, params, threshold):
+        """Test that _gradient_one_sample returns the correct values when compared to
+        calculations done by hand"""
+        cost_fn = train.Stochastic(h, vgbs)
+        sample = np.ones(dim)
+        gradient = cost_fn._gradient_one_sample(sample, params)
+
+        if threshold:
+            grad_target = np.array([18.46861628, 22.53873502, 26.60885377])
+        else:
+            grad_target = np.array([17.17157601, 20.94382262, 24.71606922])
+
+        assert np.allclose(gradient, grad_target)
+
+    @pytest.mark.parametrize("threshold", [False])
+    def test_gradient(self, vgbs, dim, params):
+        """Test that gradient returns the expected value when the VGBS class is preloaded with a
+        dataset where half of the datapoints are zeros and half of the datapoints are ones. The
+        expected result of the gradient method is then simply the average of _gradient_one_sample
+        applied to a ones vector and a zeros vector."""
+        n_samples = 10
+        zeros = np.zeros((n_samples, dim))
+        ones = np.ones((n_samples, dim))
+        samples = np.vstack([zeros, ones])
+        vgbs.add_A_init_samples(samples)
+
+        cost_fn = train.Stochastic(h, vgbs)
+        g0 = cost_fn._gradient_one_sample(zeros[0], params)
+        g1 = cost_fn._gradient_one_sample(ones[0], params)
+        grad_expected = (g0 + g1) * 0.5
+
+        grad = cost_fn.gradient(params, 2 * n_samples)
+        assert np.allclose(grad_expected, grad)
+        assert grad.shape == (dim - 1,)
+
+
+@pytest.mark.parametrize("dim", [4])
+@pytest.mark.parametrize("n_mean", [7])
+@pytest.mark.usefixtures("simple_embedding")
+class TestStochasticIntegrationPNR:
+    """Integration tests for the class ``train.Stochastic``. We consider the setting where the
+    initial adjacency matrix is the identity, which reduces to directly sampling from
+    non-interacting squeezed states without passing through an interferometer. Here,
+    the trainable parameters can only control the amount of squeezing in each mode. We then use a
+    mean squared error cost function that simply subtracts a fixed vector from an input sample and
+    squares the result."""
+
+    def identity_sampler(self, n_mean_by_mode, n_samples):
+        """Used for quickly generating samples from a diagonal adjacency matrix. This requires
+        sampling from single mode squeezed states whose photon number distribution is specified
+        by the negative binomial."""
+        ps = 1 / (1 + np.array(n_mean_by_mode))
+        np.random.seed(0)
```

Comment:
It might be good to move this into a "more global scope" (e.g. `TestStochasticIntegrationPNR.__init__`?). In more specifics, this might lead to unexpected behaviour if another method/function would be added that calls on `np.random` after seeding happened. Right now calling on `identity_sampler` leaves a state when called on by a test case and might affect other test outcomes too. Moving it to a more global scope would perhaps help with spotting that seeding happens.

**Small example**
```python
import numpy as np

class A:
    def __init__(self):
        np.random.seed(0)
        return
    
    def some_random_things(self):
        return np.random.random(55)
    
a = A()
a.some_random_things()
```
will output the same array on every call.

Reply:
Hi @trbromley, sorry for misleading you on this! It also seems that `pytest` behaves differently then simple classes making seeding more contrived than I thought so.

I think this could either: 
* be left as is (as seeding, after all, does not seem to matter between cases, this can be queried by having a look at the output of `np.random.get_state()[1][0]`)
* or a fixture could be added in `conftest.py` like
```python
@pytest.fixture
def seed():
    np.random.seed(0)
```
and making `def test_initial_cost(self, dim, n_mean, simple_embedding):` into `def test_initial_cost(self, dim, n_mean, simple_embedding, seed):`

Reply:
Thanks @antalszava, have merged in now, but adding a fixture would be good for the future!